yawl/lib/yawl/widget/Histogram.lua

```diff
@@ -0,0 +1,146 @@
+local class = require("libClass2")
+local Frame = require("yawl.widget.Frame")
+local gpu = require("component").gpu
+
+--[[
+    Ideally, Histogram will have horizontal scrolling. Though there is some contention about the direction.
+]]
+---@class Histogram:Frame
+---@field parent Frame
+---@operator call:Histogram
+---@overload fun(parent:Frame,x:number,y:number)
+local Histogram = class(Frame)
+
+---Comment
+---@return Histogram
+---@param parent Frame
+---@param x number
+---@param y number
+function Histogram:new(parent, x, y, maxColumns) --Histogramset can be string too
+    checkArg(1, parent, "table")
+    checkArg(1, maxColumns, "number", "nil")
+    local o = self.parent(parent, x, y)
+    setmetatable(o, {__index = self})
+    o._maxColumns = maxColumns
+    o._data = {}
+    o:fillChar(" ")
+    ---@cast o Histogram
+    return o
+end
+
+function Histogram:insert(value) --insert at the end
+    checkArg(1, value, "number")
+    table.insert(self._data, value)
+end
+
+function Histogram:set(index, value) --overwride the value
+    checkArg(1, index, 'number')
+    checkArg(1, value, 'number')
+    local oldValue = self._data[index]
+    if (value) then self._data[index] = value end
+    return oldValue
+end
+
+function Histogram:adjust(value, index) --change existing value
+    checkArg(1, index, 'number', 'nil')
+    checkArg(1, value, 'number')
+    index = index or #self._data --if no chosen index, change the last
+    local oldValue = self._data[index]
+    self._data[index] = oldValue + value
+    return oldValue
+end
+
+function Histogram:maxValue(value) --not the best name, basically used to control the vertical height
+    checkArg(1, value, 'number', 'nil')
+    local oldValue = self._maxValue
+    if (value) then self._maxValue = value end
+    return oldValue
+end
+
+function Histogram:clear()
+    self._data = {}
+    self:draw()
+end
+
+function Histogram:fillChar(value)
+    checkArg(1, value, 'string', 'nil')
+    local oldValue = self._fill
+    if (value) then self._fill = value end
+    return oldValue
+end
+
+function Histogram:fillForegroundColor(value)
+    checkArg(1, value, 'string', 'nil')
+    local oldValue = self._fillforegroundColor
+    if (value) then self._fillforegroundColor = value end
+    return oldValue
+end
+
+function Histogram:fillBackgroundColor(value)
+    checkArg(1, value, 'number', 'nil')
+    local oldValue = self._fillbackgroundColor
+    if (value) then self._fillbackgroundColor = value end
+    return oldValue
+end
+
+function Histogram:textForegroundColor(value)
+    checkArg(1, value, 'number', 'nil')
+    local oldValue = self._textForegroundColor
+    if (value) then self._textForegroundColor = value end
+    return oldValue
+end
+
+function Histogram:headline(value)
```

Comment:
You might want to rename this to `headlineCallback`, and copy the `Widget:callback` function

Reply:
will change later